assignment/impl/src/java/org/sakaiproject/assignment/impl/persistence/AssignmentRepositoryImpl.java

```diff
@@ -176,12 +180,41 @@ public AssignmentSubmission newSubmission(String assignmentId, Optional<String>
     }
 
     @Override
+    @SuppressWarnings("unchecked")
     public AssignmentSubmission findSubmissionForUser(String assignmentId, String userId) {
-        return (AssignmentSubmission) sessionFactory.getCurrentSession().createCriteria(AssignmentSubmission.class)
+        List<AssignmentSubmission> submissions = sessionFactory.getCurrentSession().createCriteria(AssignmentSubmission.class)
                 .add(Restrictions.eq("assignment.id", assignmentId))
                 .createAlias("submitters", "s")
                 .add(Restrictions.eq("s.submitter", userId))
-                .uniqueResult();
+                .list();
+
+        switch (submissions.size()) {
+            case 0: return null;
+            case 1: return submissions.get(0);
+            default:
+                log.info("Duplicate submissions detected for assignment {} and user {} attempting to clean", assignmentId, userId);
+                // when more than 1 was found it is considered a duplicate submission
+                // filter out user submissions and those that were submitted
+                List<AssignmentSubmission> canRemove = submissions.stream()
+                        .filter(s -> !s.getGraded() && !s.getReturned() && !s.getUserSubmission() && s.getDateSubmitted() == null)
```

Comment:
Important we only consider removing submissions that are:
- not graded
- not returned
- not a user submission
- and has never been submitted